File: dp/kadanes/cpp/max_cars_parked.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

int maxCarsParked(vector<vector<int>> array);

int maxCarsParked(vector<vector<int>> array){
    int array_size = 2*(array.size());
    vector< pair <int,bool> > parkingRegister(array_size,{0,true});
    

    for (int i = 0; i < array.size(); i++) {
        parkingRegister[2 * i] = { array[i][0], true };
        parkingRegister[2 * i + 1] = { array[i][1], false };
    }

    sort(parkingRegister.begin(), parkingRegister.end());

    int curMax=0;
    int finalMax=0;

    for (int i=0; i<array_size;++i){
        if (parkingRegister[i].second)
            curMax++;
        else{
            if (curMax > finalMax) 
                finalMax = curMax;
   
            curMax--;
        }
    }

    return finalMax;
}

int main(){

    vector<vector<int>> arr= { { 1012, 1136 },
        { 1017, 1412 },
        { 1015, 1620 } };
    
    cout<<"max cars parked ->"<< maxCarsParked(arr)<<endl;
    return 0;
}

File: dp/kadanes/python/max_cars_parked.py
def max_cars_parked(array):
    arr_len = len(array)
    parking_register = [[0,True] for i in range(2 * arr_len)]
 
    for i in range(arr_len):
        parking_register[2 * i] = [array[i][0], True]
        parking_register[2 * i + 1] = [array[i][1], False]
    
    parking_register.sort()

    curr_cnt = 0
    final_cnt = 0

    for i in range(len(parking_register)):
        if parking_register[i][1]==True:
            curr_cnt+=1
        else:
            if curr_cnt> final_cnt:
                final_cnt = curr_cnt
            curr_cnt-=1
    return final_cnt


if __name__=='__main__':
    array = [ [ 912, 1119 ],
         [ 1217, 1219 ],
         [ 1023, 1024 ],
         [1023,1027]]
    print(max_cars_parked(array))
